### PR TITLE
BC-4245 - specify kubernetes resource limits for 5 k8s entities

### PR DESCRIPTION
**Pull request by MartinSchuhmacher**
# Description
specify resource limits for ...

* api-init
* default-backend-404
* fixup-rocketchat-job
* hydra-migrate
* mongodb-query-exporter

## Links to Tickets or other pull requests
BC-4245
https://github.com/hpi-schul-cloud/schulcloud-server/pull/4225

## Deployment

* new resource request and limit values for different tasks

## Approval for review

- [ ] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### CODE REVIEW

File: ansible/roles/hydra/templates/job.yml.j2 [hunk 1]
```diff
@@ -26,6 +26,13 @@ spec:
           subPath: update.sh
         command: ['/bin/sh','-c']
         args: ['cp /update.sh /tmp/update.run.sh && chmod +x /tmp/update.run.sh && sh /tmp/update.run.sh']
+        resources:
+          limits:
+            cpu: {{ API_CPU_LIMITS|default("200m", true) }}
```
Comment (Loki-Afro) [inline ansible/roles/hydra/templates/job.yml.j2:31]: why api when this is hydra?

File: ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2 [hunk 2]
```diff
@@ -24,6 +24,13 @@ spec:
         - mountPath: /etc/nginx/conf.d/default.conf
           subPath: default.conf
           name: defaultconf
+        resources:
+          limits:
+            cpu: {{ API_CPU_LIMITS|default("200m", true) }}
```
Comment (Loki-Afro) [inline ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2:29]: why api when this is nginx?

File: ansible/roles/rocketchat/templates/fixup-job.yml.j2 [hunk 3]
```diff
@@ -26,6 +26,13 @@ spec:
               subPath: update.sh
             command: ['/bin/sh','-c']
             args: ['cp /update.sh /update.run.sh && chmod +x /update.run.sh &&./update.run.sh']
+            resources:
+              limits:
+                cpu: {{ API_CPU_LIMITS|default("300m", true) }}
```
Comment (Loki-Afro) [inline ansible/roles/rocketchat/templates/fixup-job.yml.j2:31]: why api when this is just a script?

**Review by Loki-Afro — DISMISSED** [review]

File: ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2 (hunk 2 above)
Comment (MartinSchuhmacher) [reply ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2:29]: which other env would fit here? there is only `ETHERPA_NGINX_CPU_LIMITS` and so on, but that does not seem to fit naming wise

Comment (Loki-Afro) [reply ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2:29]: just hardcode something for that one would be okay i think

Comment (mamutmk5) [reply ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2:29]: I'm opsied to hardcode this. Create an new var for it and set an default.
Why not hardcode it, for prod we have set the limit and request for memory to the same value. 
For the dev environment there is a range between request and limit for memory.

File: ansible/roles/hydra/templates/job.yml.j2 [hunk 4]
```diff
@@ -26,6 +26,13 @@ spec:
           subPath: update.sh
         command: ['/bin/sh','-c']
         args: ['cp /update.sh /tmp/update.run.sh && chmod +x /tmp/update.run.sh && sh /tmp/update.run.sh']
+        resources:
+          limits:
+            cpu: {{ HYDRA_CPU_MAX|default("200m", true) }}
+            memory: {{ HYDRA_MEM_MAX|default("20Mi", true) }}
+          requests:
+            cpu: {{ HYDRA_CPU_MIN|default("100m", true) }}
```
Comment (mamutmk5) [inline ansible/roles/hydra/templates/job.yml.j2:34]: Short hint, don't set the CPU Time Limit for an job to less then One Core.
The value for CPU is like 1/1000 second.
A value from 200 mean that the job get 200ms cpu time from one Core at a second.
If the limit is set to 1000 it's mean that the job cann ocupie an cpu core copleat at a second for each second.

Comment (MartinSchuhmacher) [reply ansible/roles/hydra/templates/job.yml.j2:34]: thx for the explanation 👍  will adjust here

File: ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2 (hunk 2 above)
Comment (Loki-Afro) [reply ansible/roles/ingress/templates/default-backend-404-deployment.yml.j2:29]: even though it is unlikely to change during my time of existence on this very earth, do what you want :D

File: ansible/roles/mongodb_query_exporter/defaults/main.yaml [hunk 5]
```diff
@@ -30,3 +30,10 @@ mongodb_query_exporter_values:
       defaultCache: 0
     aggregations:
     {{ mongodb_query_exporter_aggregations }}
+  resources:
+    limits:
+      cpu: 1000m
+      memory: 50Mi
+    requests:
+      cpu: 100m
+      memory: 30Mi
```
Comment (mamutmk5) [inline ansible/roles/mongodb_query_exporter/defaults/main.yaml:39]: For the developer enviorment it's fine to set the memory Limit and request to an different values for an Deployment.
For production and referenc systems the memory limit and request shoud be the same., so an config variable for it where be greate. End thing about a littel extra memory for more querys to run, like 128Mi of memory or so for Production.

File: ansible/roles/rocketchat/templates/fixup-job.yml.j2 [hunk 6]
```diff
@@ -26,6 +26,13 @@ spec:
               subPath: update.sh
             command: ['/bin/sh','-c']
             args: ['cp /update.sh /update.run.sh && chmod +x /update.run.sh &&./update.run.sh']
+            resources:
+              limits:
+                cpu: "1000m"
+                memory: "100Mi"
+              requests:
+                cpu: "100m"
+                memory: "50Mi"
```
Comment (mamutmk5) [inline ansible/roles/rocketchat/templates/fixup-job.yml.j2:35]: Also at this point make the memory configurable and set for production and referenc the limit and request to the same value.

**Review by mamutmk5 — CHANGES_REQUESTED** [review]

**Review by mamutmk5 — APPROVED** [review]

**Pull request merged**